hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java

```diff
@@ -73,9 +75,13 @@ public class ClientClusterServiceImpl implements ClientClusterService {
             new AtomicReference<>(new MemberListSnapshot(INITIAL_MEMBER_LIST_VERSION, new LinkedHashMap<>(), null));
     private final ConcurrentMap<UUID, MembershipListener> listeners = new ConcurrentHashMap<>();
     private final ILogger logger;
+    private final LifecycleService lifecycleService;
+    private final UUID lifecycleListenerID;
+    private volatile boolean clientConnected;
     private final Object clusterViewLock = new Object();
     //read and written under clusterViewLock
     private CountDownLatch initialListFetchedLatch = new CountDownLatch(1);
+    private boolean isFailoverSupported;
```

Comment:
Based on [this](https://stackoverflow.com/questions/32974336/is-volatile-needed-in-case-of-only-synchronized-access),  I assume both the reader and the writer should use lock. Just having the write part use lock but read part not lock, it seems that JVM does not guarantee visibility.